components/hab/src/command/pkg/download.rs

```diff
@@ -0,0 +1,439 @@
+        // Any or all of these phases should naturally fit a fork-join model
+
+        ui.begin(format!("Preparing to resolve packages and download transitive \
+                          dependendencies for {} package idents",
+                         self.idents.len()))?;
+        ui.begin(format!("Using channel {} from {}", self.channel, self.url))?;
+        ui.begin(format!("Storing in download directory {:?} ", self.download_path))?;
+
+        self.verify_and_prepare_download_directory(ui)?;
+
+        // Phase 1: Expand to fully qualified deps and TDEPS
+        let expanded_idents = self.expand_sources(ui)?;
+
+        // Phase 2: Download artifacts
+        let downloaded_artifacts = self.download_artifacts(ui, &expanded_idents)?;
+
+        Ok(downloaded_artifacts.len())
+    }
+
+    // For each source, use the builder/depot to expand it to a fully qualifed form
+    // The same call gives us the TDEPS, add those as
+    fn expand_sources<T>(&self, ui: &mut T) -> Result<HashSet<(Box<PackageIdent>, PackageTarget)>>
+        where T: UIWriter
+    {
+        let mut expanded_packages = Vec::<Package>::new();
+        let mut expanded_idents = HashSet::<(Box<PackageIdent>, PackageTarget)>::new();
+
+        // This loop should be easy to convert to a parallel map
+        for ident in &self.idents {
+            let latest = self.determine_latest_from_ident(ui, &ident.clone(), self.target);
+            if let Ok(package) = latest {
+                expanded_packages.push(package);
+            }
+        }
+
+        // Collect all the expanded deps into one structure
+        // Done separately because it's not as easy to parallelize
+        for package in expanded_packages {
+            for ident in package.tdeps {
+                expanded_idents.insert((Box::new(ident.clone()), self.target));
+            }
+            expanded_idents.insert((Box::new(package.ident.clone()), self.target));
+        }
+
+        ui.status(Status::Found,
+                  format!("{} artifacts", expanded_idents.len()))?;
+
+        Ok(expanded_idents)
+    }
+
+    fn download_artifacts<T>(&self,
+                             ui: &mut T,
+                             expanded_idents: &HashSet<(Box<PackageIdent>, PackageTarget)>)
+                             -> Result<Vec<PackageArchive>>
+        where T: UIWriter
+    {
+        let mut downloaded_artifacts = Vec::<PackageArchive>::new();
+
+        ui.status(Status::Downloading,
+                  format!("Downloading {} artifacts", expanded_idents.len()))?;
+
+        for (ident, target) in expanded_idents {
+            let archive: PackageArchive = match self.get_downloaded_archive(ui, ident, *target) {
+                Ok(v) => v,
+                Err(e) => {
+                    // Is this the right status? Or should this be a debug message?
+                    ui.status(Status::Missing,
+                              format!("Error fetching archive {} for {}: {:?}",
```

Comment:
Using the `Display` trait `{}` rather than `Debug` `{:?}`, should give more user-friendly error messages
```suggestion
                              format!("Error fetching archive {} for {}: {}",
```